### ISSUE TITLE
State gaps in connectionState algorithm

### ISSUE THREAD

**Issue by jan-ivar** [w3c/webrtc-pc#2827]
There are a gaps in the [RTCPeerConnectionStates](https://w3c.github.io/webrtc-pc/#dom-rtcpeerconnectionstate-closed) between `"new"`, `"connecting"` and `"connected"` if:
1. All ICE transports are in `"new"` or `"closed"`, and any DTLS transport is in `"connected"`, or
2. Any ICE transports are in `"new"` or `"closed"`, and all DTLS transports are `"connected"`

These don't match any state in the table:

|Enum | Description |
-- | --
new | None of the previous states apply and all [RTCIceTransport](https://w3c.github.io/webrtc-pc/#dom-rtcicetransport)s are in the "[new](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-new)" or "[closed](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-closed)" state, and all [RTCDtlsTransport](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransport)s are in the "[new](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransportstate-new)" or "[closed](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransportstate-closed)" state, or there are no transports.
connecting | None of the previous states apply and any [RTCIceTransport](https://w3c.github.io/webrtc-pc/#dom-rtcicetransport) is in the "[new](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-new)" or "[checking](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-checking)" state or any [RTCDtlsTransport](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransport) is in the "[new](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransportstate-new)" or "[connecting](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransportstate-connecting)" state.
| connected | None of the previous states apply and all [RTCIceTransport](https://w3c.github.io/webrtc-pc/#dom-rtcicetransport)s are in the "[connected](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-connected)", "[completed](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-completed)" or "[closed](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-closed)" state, and all [RTCDtlsTransport](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransport)s are in the "[connected](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransportstate-connected)" or "[closed](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransportstate-closed)" state. |

They seem impossible in practice if DTLS intrinsically depends in ICE, so this may be editorial mostly, but this isn't spelled out in the spec. Also, with n-dimensional corner cases like rollback x ICE restart x bundle renegotiation x whatever SDP negotiation mechanism may be cooked up to handle re-keying DTLS, it might be good to clean this up a bit.

One solution: reverse the order of the last two and make the latter a catch-all:
|Enum | Description |
-- | --
| connected | None of the previous states apply and all [RTCIceTransport](https://w3c.github.io/webrtc-pc/#dom-rtcicetransport)s are in the "[connected](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-connected)", "[completed](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-completed)" or "[closed](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-closed)" state, and all [RTCDtlsTransport](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransport)s are in the "[connected](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransportstate-connected)" or "[closed](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransportstate-closed)" state. |
connecting | None of the previous states apply.

This also conveniently avoids mentioning the `"checking"` ICE state which has no equivalent.

Also, after https://github.com/w3c/webrtc-pc/pull/2036 aligned `connectionState` with `iceConnectionState` it might be simpler to express one using the other, to help readers grasp their relationship, since (unlike [originally intended](https://github.com/w3c/webrtc-pc/pull/291#issuecomment-138771616)) we're likely stuck with both.

**Comment by jan-ivar** [comment]
Another problem unique to this table is the complexity added by DTLS into the logic produce sentences contain both AND and OR, which is ambiguous. E.g.

_"None of the previous states apply **_AND_** all RTCIceTransports are in the "new" or "closed" state, and all RTCDtlsTransports are in the "new" or "closed" state, **_OR_** there are no transports."_

The sanest interpretation is that lack of transports isn't meant to override previous states. It seems logical to apply the same thinking to the other states as written, since doing so creates a predictable pattern of each state deferring to the one preceding it, which seems like an intentional pattern that makes sense, and seems intended

**Comment by lgrahl** [comment]
Does WebIDL even allow to use an enum table to describe an algorithm that also relies on row order?

Regardless, I think the implementations would greatly benefit from having concrete steps in 4.4.1.3. on how to calculate the resulting connection state concretely rather than describing it in the enum table.

**Comment by jan-ivar** [comment]
> TL;DR: That's a fine modification to the algorithm if you think it makes it more readable/implementable. But the algorithm matches all possible cases, including the ones you brought up, which should return "connected".

Makes sense to me, and thanks for the proving this is editorial!

> > So... let's go back to your cases:
> > 
> > - All ICE transports are in "new" or "closed", and any DTLS transport is in "connected", or
> > - Any ICE transports are in "new" or "closed", and all DTLS transports are "connected"
> 
> In both cases, since there is an ICE transport is in the "new" state, it would trigger the rule for "connecting"

IF at least one is "new" it's "connecting" otherwise "connected", agreed.

What remains weird perhaps is this state:
- All ICE transports and DTLS transports are "closed"

...which, if it can happen outside of `pc.close()`, the spec says is "connected", when I think most people would have expected "closed".

But can this happen? DTLS can receive "close_notify" alerts, but the spec isn't clear what can may cause an ICE transport to ""shut down and ... no longer responding to STUN requests."". If it can ONLY be shutdown by `pc.close()` it might help to say that, to close the gap.

**Comment by jan-ivar** [comment]
> Does WebIDL even allow to use an enum table to describe an algorithm that also relies on row order?

WebIDL doesn't control prose. Having normative tables and algorithms referring to them to calculate a value may be unusual, but avoids repetition perhaps from keeping both algorithm steps and tables in sync.

[The PR](https://rawgit.com/jan-ivar/webrtc-pc/connectstates2/webrtc.html#rtcpeerconnectionstate-enum) doesn't remove the tables, but hopefully simplifies what they say.

**Comment by pthatcher** [comment]
If all the ICE transports and DTLS transports are "closed", then the overall state is "new".   That seems to be pretty straight forward from the rule 'all RTCIceTransports are in the "new" or "closed" state, and all RTCDtlsTransports are in the "new" or "closed" state, or there are no transports'.

That's also clear from the code "ice_new_count + ice_closed_count + dtls_new_count + dtls_closed_count == ice_total_count + dtls_total_count"

A quick reading of the Chrome code indicates that this can never happen in Chrome.  The ICE state used in the algorithm comes from here: https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/p2p/base/ice_transport_internal.h;l=249;drc=0a3ae731632f6d414e0460ab6bc0bb6e452adfda;bpv=1;bpt=1 and it never goes to "closed".  The DTLS state used in the algorithm comes from here: https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/p2p/base/dtls_transport.h;l=237;drc=0a3ae731632f6d414e0460ab6bc0bb6e452adfda;bpv=1;bpt=1 and there are 2 places that the DTLS state can go to closed: https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/p2p/base/dtls_transport.cc;l=731;drc=0a3ae731632f6d414e0460ab6bc0bb6e452adfda and https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/p2p/base/dtls_transport.cc;l=712;drc=0a3ae731632f6d414e0460ab6bc0bb6e452adfda.   So, the DTLS state could go to "closed", but no the ICE state, so the overall state could not hit this situation.
